src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs

```diff
@@ -1365,17 +1440,39 @@ private IEnumerable<IDiagnostic> GetOutputDeclarationDiagnostics(TypeSymbol assi
             var valueType = typeManager.GetTypeInfo(syntax.Value);
 
             // this type is not a property in a symbol so the semantic error visitor won't collect the errors automatically
-            if (valueType is ErrorType)
+            var diagnostics = new List<IDiagnostic>();
+            diagnostics.AddRange(valueType.GetDiagnostics());
+
+            if (assignedType is ResourceType resourceType &&
+                syntax.Type is ResourceTypeSyntax resourceTypeSyntax)
             {
-                return valueType.GetDiagnostics();
+                if (IsExtensibilityType(resourceType))
+                {
+                    diagnostics.Add(DiagnosticBuilder.ForPosition(resourceTypeSyntax).UnsupportedResourceTypeOutputType(assignedType.Name));
+                }
+
+                if (!features.ResourceTypedParamsAndOutputsEnabled)
+                {
+                    diagnostics.Add(DiagnosticBuilder.ForPosition(syntax.Span).ParamOrOutputResourceTypeUnsupported());
+                }
+
+                // As a special case of outputs, we don't want to double-up diagnostics on inferred resource types.
+                // The inference is based on another declaration in the file, and so the user should fix that instead.
+                if (resourceTypeSyntax.Type is {}
+                    && !resourceType.DeclaringNamespace.ResourceTypeProvider.HasDefinedType(resourceType.TypeReference))
+                {
+                    diagnostics.Add(DiagnosticBuilder.ForPosition(resourceTypeSyntax.Type!).ResourceTypesUnavailable(resourceType.TypeReference));
+                }
             }
 
-            if (TypeValidator.AreTypesAssignable(valueType, assignedType) == false)
+            // Avoid reporting an additional error if we failed to bind the output type.
+            if (TypeValidator.AreTypesAssignable(valueType, assignedType) == false &&
```

Comment:
>== false [](http://example.com/codeflow?start=74&length=8)

The `== false` stuff is a leftover from when this method return type was `bool?`. Can probably simplify.